### PR TITLE
EVG-18798: Enable fetching and merging test results from multiple tasks

### PR DESCRIPTION
**Pull request by julianedwards**
EVG-18798: https://jira.mongodb.org/browse/EVG-18798

This is the first of three PRs for this ticket.

- Enable the test results DB functions to accept multiple (task ID, execution) pairs for finding and fetching test results
- Maintain REST route parity with existing routes.
- Update tests
- Clean up some documentation and function signatures

I updated the tests to reflect the changes and ensure parity with the current REST API. I also verified in staging that the test results API behaves as expected with the Evergreen UI.

### CODE REVIEW

File: model/test_results.go [hunk 1]
```diff
@@ -755,64 +740,66 @@ func FindTestResults(ctx context.Context, env cedar.Environment, opts FindTestRe
...
+				execution: record.Info.Execution,
 			}
-			displayTaskMap[dt] = append(displayTaskMap[dt], result.FailedTestsSample...)
+			displayTaskMap[dt] = append(displayTaskMap[dt], record.FailedTestsSample...)
 		}
 		et := taskExecutionPair{
-			taskID:    result.Info.TaskID,
-			execution: result.Info.Execution,
+			taskID:    record.Info.TaskID,
+			execution: record.Info.Execution,
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/test_results.go:796]: It doesn't, it's a bug. Our routes can no longer support fetching by display task IDs but I need to support this for backwards compatibility until the changes are made on the front-end.

File: model/test_results.go [hunk 2]
```diff
@@ -1015,15 +979,21 @@ func FindAndDownloadTestResults(ctx context.Context, env cedar.Environment, opts
...
 	if catcher.HasErrors() {
-		return nil, 0, catcher.Resolve()
+		return TestResultsStats{}, nil, catcher.Resolve()
+	}
+
+	filteredResults, filteredCount, err := filterAndSortTestResults(ctx, env, combinedResults, filterOpts)
+	if err != nil {
+		return TestResultsStats{}, nil, errors.Wrap(err, "filtering and sorting test results")
 	}
+	stats.FilteredCount = &filteredCount
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/test_results.go:989]: `FilteredCount` is the total number of test results matching the given filter parameters, if it's paginated we can't use the length of the filtered results as there may be more results in a subsequent page (see L1018).

File: rest/data/test_results.go [hunk 3]
```diff
@@ -7,35 +7,24 @@ import (
...
 	"github.com/pkg/errors"
 )
 
 /////////////////////////////
 // DBConnector Implementation
 /////////////////////////////
 
-func (dbc *DBConnector) FindTestResults(ctx context.Context, opts TestResultsOptions) (*model.APITestResults, error) {
-	apiStats, err := dbc.GetTestResultsStats(ctx, opts)
+func (dbc *DBConnector) FindTestResults(ctx context.Context, taskOpts []TestResultsTaskOptions, filterOpts *TestResultsFilterAndSortOptions) (*model.APITestResults, error) {
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply rest/data/test_results.go:17]: That will be in the next PR when I make the new routes :)

File: rest/data/test_results.go [hunk 4]
```diff
@@ -181,46 +158,29 @@ func extractFailedTestResultsSample(results ...dbModel.TestResults) []string {
...
-		Find:          convertToDBFindTestResultsOptions(opts),
-		FilterAndSort: filterAndSort,
+	return &dbModel.TestResultsFilterAndSortOptions{
+		TestName:     opts.TestName,
+		Statuses:     opts.Statuses,
+		GroupID:      opts.GroupID,
+		SortBy:       dbModel.TestResultsSortBy(opts.SortBy),
+		SortOrderDSC: opts.SortOrderDSC,
+		Limit:        opts.Limit,
+		Page:         opts.Page,
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply rest/data/test_results.go:184]: Good catch!

File: rest/service.go [hunk 5]
```diff
@@ -318,14 +318,12 @@ func (s *Service) addRoutes() {
 	s.app.AddRoute("/buildlogger/test_name/{task_id}/{test_name}/meta").Version(1).Get().Wrap(evgAuthReadLogByTaskID).RouteHandler(makeGetLogMetaByTestName(s.sc))
 	s.app.AddRoute("/buildlogger/test_name/{task_id}/{test_name}/group/{group_id}").Version(1).Get().Wrap(evgAuthReadLogByTaskID).RouteHandler(makeGetLogGroupByTestName(s.sc))
 
-	s.app.AddRoute("/test_results/filtered_samples").Version(1).Get().RouteHandler(makeGetTestResultsFilteredSamples(s.sc))
+	// TODO (EVG-18798): Remove these once Spruce and Evergreen are
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply rest/service.go:321]: All of these routes will be replaced with the new routes that will accept multiple (task ID, execution) pairs. But that will happen after I create the routes, update timber, and then update Evergreen/Spruce.

File: rest/test_results_routes.go [hunk 6]
```diff
@@ -101,12 +105,14 @@ func (h *testResultsGetByTaskIDHandler) Parse(ctx context.Context, r *http.Reque
...
+		h.filterOpts.SortOrderDSC = true
 	}
 	if baseTaskID != "" {
-		h.opts.FilterAndSort.BaseResults = &data.TestResultsOptions{
-			TaskID:      baseTaskID,
-			DisplayTask: h.opts.DisplayTask,
+		h.filterOpts.BaseResults = []data.TestResultsTaskOptions{
+			{
+				TaskID:      baseTaskID,
+				DisplayTask: h.taskOpts.DisplayTask,
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply rest/test_results_routes.go:114]: It fetches the most recent execution of the task that Cedar has, so we pass in a `nil` execution.

**Review by ybrill — APPROVED** [review]

**Pull request merged**